File: src/system/history.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

const HISTORY_PATH = path.join(os.homedir(), '.config', 'anpan', 'history.json')
const MAX_ENTRIES = 50

export function loadHistory(): string[] {
  try {
    const data: unknown = JSON.parse(fs.readFileSync(HISTORY_PATH, 'utf8'))
    return Array.isArray(data) ? data.filter((e): e is string => typeof e === 'string') : []
  } catch {
    return []
  }
}

export function addToHistory(url: string): string[] {
  const updated = [url, ...loadHistory().filter(e => e !== url)].slice(0, MAX_ENTRIES)
  try {
    fs.mkdirSync(path.dirname(HISTORY_PATH), {recursive: true})
    fs.writeFileSync(HISTORY_PATH, `${JSON.stringify(updated, null, 2)}\n`)
  } catch {}
  return updated
}

File: internal/system/history.go
package system

import (
	"encoding/json"
	"os"
	"path/filepath"
)

const maxHistoryEntries = 50

func historyPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".config", "anpan", "history.json")
}

func LoadHistory() []string {
	data, err := os.ReadFile(historyPath())
	if err != nil {
		return []string{}
	}
	var res []string
	if err := json.Unmarshal(data, &res); err == nil {
		return res
	}
	return []string{}
}

func AddToHistory(url string) []string {
	history := LoadHistory()
	var updated []string
	updated = append(updated, url)
	for _, item := range history {
		if item != url {
			updated = append(updated, item)
		}
	}
	if len(updated) > maxHistoryEntries {
		updated = updated[:maxHistoryEntries]
	}

	p := historyPath()
	_ = os.MkdirAll(filepath.Dir(p), 0o755)
	if data, err := json.MarshalIndent(updated, "", "  "); err == nil {
		_ = os.WriteFile(p, append(data, '\n'), 0o644)
	}
	return updated
}
